mangaki/mangaki/utils/wals.py

```diff
@@ -0,0 +1,130 @@
+from collections import defaultdict, Counter
+from mangaki.utils.chrono import Chrono
+import numpy as np
+import pickle
+
+import tensorflow as tf
+from tensorflow.contrib.factorization.python.ops import factorization_ops
+from tensorflow.python.framework import sparse_tensor
+
+sess = tf.InteractiveSession()
+
+def simple_train(model, inp, num_iterations):
+  """Helper function to train model on inp for num_iterations."""
+  row_update_op = model.update_row_factors(sp_input=inp)[1]
+  col_update_op = model.update_col_factors(sp_input=inp)[1]
+
+  model.initialize_op.run()
+  model.worker_init.run()
+  for _ in range(num_iterations):
+    model.row_update_prep_gramian_op.run()
+    model.initialize_row_update_op.run()
+    row_update_op.run()
+    model.col_update_prep_gramian_op.run()
+    model.initialize_col_update_op.run()
+    col_update_op.run()
+
+class MangakiWALS(object):
+    M = None
+    U = None
+    VT = None
+    def __init__(self, NB_COMPONENTS=10, NB_ITERATIONS=10, LAMBDA=0.1):
+        self.NB_COMPONENTS = NB_COMPONENTS
+        self.NB_ITERATIONS = NB_ITERATIONS
+        self.LAMBDA = LAMBDA
+        self.chrono = Chrono(True)
+
+    def save(self, filename):
+        with open(filename, 'wb') as f:
+            pickle.dump(self, f)
+
+    def load(self, filename):
+        with open(filename, 'rb') as f:
+            backup = pickle.load(f)
+        self.M = backup.M
+        self.U = backup.U
+        self.VT = backup.VT
+
+    def set_parameters(self, nb_users, nb_works):
+        self.nb_users = nb_users
+        self.nb_works = nb_works
+
+    def make_matrix(self, X, y):
+        matrix = defaultdict(dict)
+        nb_ratings = Counter()
+        means = np.zeros((self.nb_users,))
+        users = set()
+        for (user, work), rating in zip(X, y):
+            matrix[(user, work)] = rating
+            means[user] += rating
+            nb_ratings[user] += 1.
+            users.add(user)
+        for user in users:
+            means[user] /= nb_ratings[user]
+        indices = []
+        values = []
+        for (user, work) in X:
+            matrix[(user, work)] -= means[user]
+            indices.append((user, work))
+            values.append(matrix[(user, work)])
+        print(indices[:5])
+        return indices, values, means
+        # return matrix, means
+
+    def fit_user(self, user, matrix):
+        Ru = np.array(list(matrix[user].values()), ndmin=2).T
```

Comment:
Please comment what `Ru`, `Vu`, `Gu` are, at least qualitatively (add doc in the function).